rfcs/unboxed-types.md

```diff
@@ -0,0 +1,2036 @@
+    word      | word_rep       | false       | true
+```
+
+Just as `immediate <= value` works, `immediate <= word` works:
+`immediate` is really the combination of `value` (with its
+gc-friendliness) and `word` (with its gc-ignorableness).
+
+However, we hold off on adding `immediate <= word` to our type
+system for now, because it adds complication without anyone
+asking for it.
+
+Here are some complications introduced:
+
+* It is very convenient to reduce computing
+`la ⊓ lb` to a constraint `rep(la) = rep(lb)`. In order to support `value ⊓ word
+= immediate`, we would thus drop the `value_rep` representation, setting
+`rep(value) = word_rep`.
+
+* The `upper_bound` function we use in type inference then becomes impossible to
+define: neither `value` nor `word` is an upper bound for the other, and so
+`upper_bound(word_rep)` has no definition.  We thus have to design a different
+type inference algorithm than the one described above, likely tracking
+representation, gc-friendliness, and gc-ignorability separately. This would not
+be all that hard, but it's definitely more complicated than the algorithm
+proposed here.
+
+* We sometimes might infer `immediate` even when no function
+nearby has anything of layout `immediate`:
+
+    ```ocaml
+    val f1 : ('a : value). 'a -> ...
+    val f2 : ('a : word). 'a -> ...
+
+    let f3 x = (f1 x, f2 x)
+    ```
+
+    Here, we need to infer that the `x` in `f3` has layout
+    `immediate`, even though neither `f1` nor `f2` mentions
+    `immediate`. This might be unexpected by users.
+
+## Further discussion: mutability within unboxed records
+
+This proposal forbids mutability within an unboxed record (unless that
+record is itself contained within a boxed record). This decision is made to
+reduce the possibility of confusion and bugs, not because mutability within
+unboxed records is unsound. Here, we consider some reasons that push against the
+decision above and include some examples illustrating the challenges.
+
+Reasons we might want mutability within unboxed types:
+
+* The designer of a type may choose to have some fields be mutable while
+  other fields would not be. For example, our `#mut_point` type allows the
+  location of the point to change, while the identity of the point must
+  remain the same.
+
+* OCaml programmers expect to be able to update their mutable fields, and it
+  may cause confusion if a programmer cannot, say, update the `x` field of a
+  `#mut_point`.
+
+The reason we decided not to allow mutability within unboxed types is because of
+examples like this:
+
+```ocaml
+type mut_point = { mutable x : #int32; mutable y : #int32; identity : int }
+let () =
+  let p : #mut_point = #{ x = #4l; y = #5l; identity = 0 } in
+  let foo () = print_uint32 p.x in
+  foo ();
+  p.x <- #10l;
+  foo ()
+```
+
+Because the closure for `foo` captures that unboxed record `p` (by copying `p`),
+this code prints 4 twice, rather confusingly. Yet the current design does not
```

Comment:
Let's come back to this example, after having come to the conclusion that what happens here is eventually unrelated to the other examples below it.

I think the surprising choice here is to copy by value when the user could have written more explicitly:  
```ocaml
  let x = p.x in
  let foo () = print_uint32 x in
```
if they wanted a copy.

Besides interfering with mutations, you note elsewhere that this copy-on-capture rule can also be surprising because of the cost of hidden copies (and AFAIU this is the only place where the copy is _really_ hidden by the syntax).

I would expect instead:
- `p` is captured by reference.
- The program is rejected if the reference to the stack can escape (e.g. by forcing to annotate the function with a `local` keyword).
- Mutations on the stack are allowed.

If not possible, I would rather those captures and hidden copies to be forbidden (and mutations on the stack allowed). (To be able to consider this, it was important for me to realize that the subsequent examples are unrelated to this one.)

In general in type systems, the interaction between kinds and closures is a delicate aspect that is not arbitrary; and I worry that this proposal commits to an easy-to-implement but not-so-useful solution.

Reply:
> (and AFAIU this is the only place where the copy is _really_ hidden by the syntax)

I'm not sure what's the problem with that. There's no real limit to how much a closure can capture. Even if you write `fun x -> foo x`, it may seem that you only capture `foo` but if `foo` gets expanded before closure conversion (for instance, by the `[@local]` optimisation) then you could have an arbitrary large number of variables captured.
I'm not saying we should encourage that, and I have actually proposed a restriction to the `[@local]` optimisation that would prevent that from occurring, but capturing the whole layout of an unboxed record doesn't seem particularly surprising to me. 

> * `p` is captured by reference.

I think that the "reference" keyword is a red flag. We don't have any concept of passing by reference in OCaml (except for the `ref` type), and I don't think we need to introduce it to make sense of unboxed records.
In C terms, an unboxed record is basically a `struct`. When you pass it as argument, you pass it by value, copying the whole of it. If you want to pass it by reference, there's already a natural way to do so: use the boxed version (which is basically a pointer to a `struct`).
The fact that this applies to closure capture too shouldn't be considered particularly harmful, in my opinion. And it's relatively easy for an optimising compiler to trim the closure if needed.

Reply:
If anything, C gives a bad comparison precisely because it makes it easy to silently spill it on the stack and mutate it by reference. I agree that there can be value in more clearly separating the notions of unboxing and stack allocation than in those languages.